src/UCTNode.cpp

```diff
@@ -412,133 +349,79 @@ UCTNode* UCTNode::uct_select_child(int color) {
+class NodeComp : public std::binary_function<std::unique_ptr<UCTNode>&,
+                                             std::unique_ptr<UCTNode>&, bool> {
 public:
-    NodeComp() = default;
+    NodeComp(int color) : m_color(color) {};
+
     // winrate, visits, score, child
     //        0,     1,     2,     3
 
-    bool operator()(const UCTNode::sortnode_t a, const UCTNode::sortnode_t b) {
-        // One node has visits, the other does not
-        if (!std::get<1>(a) && std::get<1>(b)) {
-            return false;
-        }
-
-        if (!std::get<1>(b) && std::get<1>(a)) {
-            return true;
-        }
-
-        // Neither has visits, sort on prior score
-        if (!std::get<1>(a) && !std::get<1>(b)) {
-            return std::get<2>(a) > std::get<2>(b);
+    bool operator()(const std::unique_ptr<UCTNode>& a,
+                    const std::unique_ptr<UCTNode>& b) {
+        // if visits are not same, sort on visits
+        if (a->get_visits() != b->get_visits()) {
+            return a->get_visits() < b->get_visits();
         }
 
-        // Both have visits, but the same amount, prefer winrate
-        if (std::get<1>(a) == std::get<1>(b)) {
-            return std::get<0>(a) > std::get<0>(b);
+        // neither has visits, sort on prior score
+        if (a->get_visits() == 0) {
+            return a->get_score() < b->get_score();
         }
 
-        // Both have different visits, prefer greater visits
-        return std::get<1>(a) > std::get<1>(b);
+        // both have same non-zero numbor of visits
+        return a->get_eval(m_color) < b->get_eval(m_color);
     }
+private:
+    int m_color;
+
 };
 
 void UCTNode::sort_root_children(int color) {
     LOCK(get_mutex(), lock);
-    auto tmp = std::vector<sortnode_t>{};
-
-    auto child = m_firstchild;
-    while (child != nullptr) {
-        auto visits = child->get_visits();
-        auto score = child->get_score();
-        if (visits) {
-            auto winrate = child->get_eval(color);
-            tmp.emplace_back(winrate, visits, score, child);
-        } else {
-            tmp.emplace_back(0.0f, 0, score, child);
-        }
-        child = child->m_nextsibling;
-    }
-
-    // reverse sort, because list reconstruction is backwards
-    std::stable_sort(rbegin(tmp), rend(tmp), NodeComp());
-
-    m_firstchild = nullptr;
-
-    for (auto& sortnode : tmp) {
-        link_child(std::get<3>(sortnode));
-    }
-}
-
-/**
- * Helper function to get a sortnode_t
- * eval is set to 0 if no visits instead of first-play-urgency
- */
-UCTNode::sortnode_t get_sortnode(int color, UCTNode* child) {
-    auto visits = child->get_visits();
-    return UCTNode::sortnode_t(
-        visits == 0 ? 0.0f : child->get_eval(color),
-        visits,
-        child->get_score(),
-        child);
+    std::stable_sort(begin(m_children), end(m_children),
+                     NodeComp(color));
+    std::reverse(begin(m_children), end(m_children));
```

Comment:
This can be verified by the command in comment number two

Reply:
I think @gcp mentioned it before, but why not reverse the sort order in `NodeComp` so that `std::reverse` could be removed?

Reply:
It doesn't sort the same because of stable.

2 1a 1b 5 3
sort_stable => 1a 1b 2 3 5
reverse => 5 3 2 1b 1a

sort_stable (reverse compare) => 5 3 2 1a 1b

all the elements that are equal will be in the wrong order.

I thought that by using rbegin and rend I could fix this but I haven't made it work